sympy/core/evalf.py

```diff
@@ -1260,10 +1260,12 @@ def hypsum(expr: 'Expr', n: 'Symbol', start: int, prec: int) -> mpf:
     if h < 0:
         raise ValueError("Sum diverges like (n!)^%i" % (-h))
 
-    term = expr.subs(n, 0)
-    if not term.is_Rational:
+    eterm = expr.subs(n, 0)
+    if not eterm.is_Rational:
         raise NotImplementedError("Non rational term functionality is not implemented.")
 
+    term: Rational = eterm # type: ignore
+
```

Comment:
I think that `is_Rational` is not useful for type guarding.
Maybe [PEP-647](https://peps.python.org/pep-0647/) may help, but I don't think that it is possible, or at least easy how to make `obj.is_Rational` work.

However, the suggestion is that could we just use `isinstance(term, Rational)` if the performance changes of using `isinstance` is asymptotically not significant for the case.

Reply:
This example could use `isinstance` but there are many other cases like this throughout the codebase and I am not sure how best to handle them in general. I have only changed these when mypy reports an error. Currently mypy checks this function but infers that `subs` returns `Any`. The change in this PR means that it infers that `subs` returns `Expr` and then it complains that `Expr` does not have a `.p` attribute.